titus-server-master/src/main/java/com/netflix/titus/master/scheduler/constraint/V3ZoneBalancedFitnessCalculator.java

```diff
@@ -0,0 +1,66 @@
+package com.netflix.titus.master.scheduler.constraint;
+
+import java.util.Map;
+
+import com.netflix.fenzo.TaskRequest;
+import com.netflix.fenzo.TaskTrackerState;
+import com.netflix.fenzo.VMTaskFitnessCalculator;
+import com.netflix.fenzo.VirtualMachineCurrentState;
+import com.netflix.titus.master.jobmanager.service.common.V3QueueableTask;
+import com.netflix.titus.master.scheduler.SchedulerUtils;
+
+public class V3ZoneBalancedFitnessCalculator implements VMTaskFitnessCalculator {
+
+    private final double NOT_MATCHING = 0.01;
+    private final double MATCHING = 1.0;
+
+    private final TaskCache taskCache;
+    private final int expectedValues;
+    private final String zoneAttributeName;
+
+    public V3ZoneBalancedFitnessCalculator(TaskCache taskCache, int expectedValues, String zoneAttributeName) {
+        this.taskCache = taskCache;
+        this.expectedValues = expectedValues;
+        this.zoneAttributeName = zoneAttributeName;
+    }
+
+    @Override
+    public String getName() {
+        return "V3ZoneBalancedFitnessCalculator";
+    }
+
+    @Override
+    public double calculateFitness(TaskRequest taskRequest, VirtualMachineCurrentState targetVM, TaskTrackerState taskTrackerState) {
+        // Ignore the constraint for non-V3 tasks.
+        if (!(taskRequest instanceof V3QueueableTask)) {
+            return MATCHING;
+        }
+
+        String targetZoneId = SchedulerUtils.getAttributeValueOrEmptyString(targetVM, zoneAttributeName);
+        if (targetZoneId.isEmpty()) {
```

Comment:
fail precondition or log something, as we don't expect agents to not have a `zoneAttribute`?